### ISSUE TITLE
[rocJITsu][gfx1250] PyTorch FP8 E4M3-to-float conversion returns incorrect values

### ISSUE THREAD

**Issue by Yu-Zhewen** [ROCm/rocm-systems#10760]
# [rocJITsu][gfx1250] PyTorch FP8 E4M3-to-float conversion returns incorrect values

## Summary

On pinned rocJITsu commit `08ae38600aa96bc29fde7701ef2d17240d4f1945`
with the `gfx1250_mi455x.json` configuration, converting a GPU
`torch.float8_e4m3fn` tensor to `torch.float32` returns incorrect values.

The same PyTorch package and input produce the expected values under MI450
FFM. Converting BF16 to FP8 on rocJITsu produces the correct raw FP8 bytes, so
the failure is specific to conversion from FP8.

## Environment

- rocJITsu commit: `08ae38600aa96bc29fde7701ef2d17240d4f1945`
- Configuration: `emulation/rocjitsu/configs/gfx1250_mi455x.json`
- PyTorch: `2.13.0+rocm10.1.0a20260812`
- HIP runtime reported by PyTorch: `7.16.26315`
- GPU target: `gfx1250`

## Minimal reproducer

```python
import torch

values = torch.tensor(
    [-4.0, -2.0, -1.0, -0.5, -0.25, 0.0, 0.25, 0.5, 1.0, 2.0, 4.0, 8.0],
    dtype=torch.float32,
).to(torch.float8_e4m3fn)

# Construct FP8 on the CPU and copy its raw representation to the GPU.
device_values = values.to("cuda")

print("raw:", device_values.view(torch.uint8).cpu().tolist())
print("expected:", values.float().tolist())
print("actual:", device_values.float().cpu().tolist())
```

Run:

```bash
ROCJITSU=/path/to/rocjitsu-build/tools/rocjitsu/rocjitsu
CONFIG=/path/to/rocm-systems/emulation/rocjitsu/configs/gfx1250_mi455x.json

"${ROCJITSU}" --daemon \
  --config "${CONFIG}" \
  -- python repro.py
```

## Actual result

```text
raw: [200, 192, 184, 176, 168, 0, 40, 48, 56, 64, 72, 80]
expected: [-4.0, -2.0, -1.0, -0.5, -0.25, 0.0, 0.25, 0.5, 1.0, 2.0, 4.0, 8.0]
actual: [-0.0625, -0.0625, -0.03125, -0.03125, -0.03125, 0.0, 0.03125, 0.03125, 0.03125, 0.0625, 0.0625, 0.0625]
```

The raw bytes on the GPU exactly match the source FP8 bytes, but converting
them to FP32 returns incorrect values.

## Expected result

```text
actual: [-4.0, -2.0, -1.0, -0.5, -0.25, 0.0, 0.25, 0.5, 1.0, 2.0, 4.0, 8.0]
```

The same script produces this expected result under MI450 FFM.

## Additional isolation

- GPU BF16-to-FP8 conversion produces the expected raw E4M3 bytes under
  rocJITsu.
- A minimal Triton kernel loading the same FP8 bytes and converting them with
  `tl.load(...).to(tl.float32)` produces the expected values under rocJITsu.
- Minimal HIP kernels using scalar `__hip_fp8_e4m3` and packed
  `__hip_fp8x2_e4m3` conversions both produce the expected values under
  rocJITsu.
- The failure is therefore specific to the GPU conversion path exercised by
  PyTorch's cast kernel, rather than FP8 storage or host-to-device copying.

**Comment by sjain-stanford** [comment]
Reproduced on recent `develop` HEAD @ `97ab4d1f880b1814f736de11453b525bed21a250`.

Using the issue’s PyTorch/ROCm nightly versions:

```text
torch: 2.13.0+rocm10.1.0a20260812
hip:   7.16.26315
raw:      [200, 192, 184, 176, 168, 0, 40, 48, 56, 64, 72, 80]
expected: [-4, -2, -1, -0.5, -0.25, 0, 0.25, 0.5, 1, 2, 4, 8]
actual:   [-0.0625, -0.0625, -0.03125, -0.03125, -0.03125, 0,
            0.03125, 0.03125, 0.03125, 0.0625, 0.0625, 0.0625]
```

This exactly matches [issue #10760](https://github.com/ROCm/rocm-systems/issues/10760). It also reproduces with `RJ_FORCE_SCALAR=1`, so it affects both scalar and SIMD execution paths.

Root cause
----------

This is not a `v_cvt_f32_fp8` bug. PyTorch implements this conversion with integer bit manipulation. The first incorrect intermediate comes from:

```text
encoding: d5268004 02010904
v_sub_nc_u32 v4, v4, 4 clamp
```

The encoding has `clamp=1`. For FP8 byte `0xc8` (`-4`):

```text
nonsign       = 0x48000000
clz(nonsign)  = 1
desired shift = saturating_sub(1, 4) = 0
rocJITsu      = wrapping_sub(1, 4)   = 0xfffffffd
```

That wrapped `-3` feeds exponent reconstruction, producing `0x3d800000`, or `0.0625`, rather than `0x40800000`, or `4.0`. The sign is subsequently restored correctly.

rocJITsu exposes the VOP3 clamp field in [machine_insts.h](https://github.com/ROCm/rocm-systems/blob/97ab4d1f880b1814f736de11453b525bed21a250/emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/isa/arch/amdgpu/generated/cdna5/machine_insts.h#L66), but:

- The scalar generator performs unconditional wrapping arithmetic in [vector_alu.py](https://github.com/ROCm/rocm-systems/blob/97ab4d1f880b1814f736de11453b525bed21a250/emulation/rocjitsu/lib/python/amdisa/codegen/execute/vector_alu.py#L808).
- The SIMD mapping similarly uses plain `a - b` in [simd_codegen.py](https://github.com/ROCm/rocm-systems/blob/97ab4d1f880b1814f736de11453b525bed21a250/emulation/rocjitsu/lib/python/amdisa/codegen/execute/simd_codegen.py#L1818).
- The SIMD helper explicitly assumes integer clamp is never requested in [simd_glue.h](https://github.com/ROCm/rocm-systems/blob/97ab4d1f880b1814f736de11453b525bed21a250/emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/isa/arch/amdgpu/shared/simd_glue.h#L1674).

Proposed robust fix
-------------------

Implement ISA-defined integer saturation in the code generator, rather than special-casing FP8 or PyTorch:

- Classify VOP3 integer arithmetic opcodes that define clamp/saturation.
- For unsigned subtraction with clamp:
  `a < b ? 0 : a - b`.
- For unsigned addition, saturate overflow to `UINT_MAX`.
- Handle reverse subtraction and signed/16-bit variants with their corresponding limits.
- Make the scalar generated implementation authoritative.
- Initially make the SIMD path fall back to scalar whenever a saturation-capable integer instruction has `clamp=1`. This is the safest correctness fix.
- Optionally add properly masked SIMD saturation afterward, once boundary-tested.
- Regenerate generated execution files; do not patch `generated/cdna5/vop3_exec_alu.cpp` directly.

Tests should include:

- An execution-harness regression using the exact failing encoding and inputs below, equal to, and above `4`.
- Clamp-on versus clamp-off behavior.
- Unsigned add/sub/reverse-sub boundary cases and signed/16-bit equivalents supported by the ISA.
- Both normal SIMD execution and forced-scalar execution.
- A small compiled HIP integration test implementing PyTorch’s E4M3 conversion algorithm, covering zero, subnormals, normals, maximum finite values, and NaNs.

**Comment by RattataKing** [comment]
Adding two pieces of information beyond the existing source-level diagnosis:
a framework-minimized confirmation and the shared-fix relationship with
#10762. Checked against rocJITsu `develop` at
`79f87b8238f04a255707636719ba0a9f525753dd`.

## Detailed root cause

A header-only HIP probe that includes C10's `Float8_e4m3fn.h`, without
TensorIterator, Triton, or tensor storage, reproduces the value collapse in
both normal and `RJ_FORCE_SCALAR=1` execution. The generated decode contains:

```asm
v_clz_i32_u32       v3, v2
v_min_u32           v3, 32, v3
v_sub_nc_u32_e64    v3, v3, 4 clamp
```

C10 needs `max(clz(nonsign)-4, 0)`. For raw E4M3 `0x28`, `nonsign` is
`0x28000000` and CLZ is 2. Correct clamp produces shift 0; rocJITsu produces
`2-4 = 0xfffffffe`. The later variable shift uses the low five bits, selecting
shift 30 and generating FP32 bits `0x3d000000`, or 0.03125, instead of 0.25.

The implementation omission is shared with #10762:

- `execute_v_sub_nc_u32_vop3()` performs wrapping `a-b` in scalar and SIMD
  paths without reading `clamp`;
- `execute_v_subrev_nc_u32_vop3()` performs wrapping `b-a` with the same
  omission;
- the generic SIMD helper assumes integer clamp is unused.

The six affected Triton `test_cast` cases use a PyTorch GPU cast as their
reference. That reference runs this faulty C10 sequence; Triton's direct FP8
widening is the correct side of those comparisons.

## Suggested fix and regression coverage

Fix #10760 and #10762 together. With `clamp=1`, use unsigned saturating
subtraction in both directions; with `clamp=0`, retain wrapping arithmetic.
Implement this in the authoritative generator/semantic layer and regenerate
checked-in sources. Make SIMD clamp-aware or fall back to the corrected scalar
path for clamped operations.

Regression coverage should include:

- normal and reversed subtraction below/equal/above the underflow boundary;
- clamp on/off;
- scalar and SIMD paths;
- literal, SGPR, VGPR, and legal DPP forms;
- the header-only C10 E4M3 conversion over zero, subnormals, normals, maximum
  finite values, and NaNs;
- the six E4M3-to-BF16/FP16/FP32 workload cases;
- an audit of other integer VOP3 operations routed through the same
  clamp-agnostic SIMD helper.

**Comment by atgutier** [comment]
An agent audited this and the root cause is correct, but the fix may need some refinement:

[MAJOR] The draft omits gfx1250’s clamped 64-bit add/sub instructions.

The new classification only covers i16/u16/i32/u32. At PR head 33a8bd0f, these still perform unconditional wrapping arithmetic:

- V_ADD_NC_U64
- V_SUB_NC_U64
- Their DPP paths

That is inconsistent with LLVM’s instruction metadata: VOP_I64_I64_I64_ARITH enables integer clamp in llvm-project/llvm/lib/Target/AMDGPU/SIInstrInfo.td:3140, and LLVM accepts explicit clamped gfx1250 encodings in rocm-dev/llvm-project/llvm/test/MC/AMDGPU/gfx1250_asm_vop2.s:275.

The implementation should add u64 support:

- Addition: saturate when lhs > UINT64_MAX - rhs.
- Subtraction: saturate to zero when lhs < rhs.
- Preserve ordinary modulo arithmetic with clamp=0.
- Test direct and DPP execution.

[MAJOR] The proposal is not yet an exhaustive “integer CLAMP” fix.

Several other instructions classified by LLVM as integer-clamp operations still ignore inst_.clamp, including:

- V_MAD_I32_I24, V_MAD_U32_U24
- V_MAD_I16, V_MAD_U16
- V_SAD_U8, V_SAD_HI_U8, V_SAD_U16
- V_MSAD_U8
- V_QSAD_PK_U16_U8
- V_MQSAD_PK_U16_U8, V_MQSAD_U32_U8

LLVM explicitly groups several of these under “Integer Clamp Patterns” in llvm-project/llvm/lib/Target/AMDGPU/VOP3Instructions.td:2068. Their rocJITsu scalar bodies still wrap, and their SIMD paths remain active with clamp=1.

These need instruction-specific treatment: CLAMP generally applies to the final arithmetic result, so extending the current generic nested ADD/SUB lowering mechanically could incorrectly saturate an intermediate expression. Packed SAD variants also require their exact component saturation rules to be verified.
